styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -0,0 +1,364 @@
+    cache.put(counterId, newSnapshot);
+    return newSnapshot;
+  }
+
+  /**
+   * Must be called within a TransactionCallable. (?)
+   *
+   * <p>Augments the transaction with operations to persist the given limit in Datastore. So long as
+   * there has been no preceding successful updateLimit operation, no limit is applied in
+   * updateCounter operations on this counter.
+   */
+  public void updateLimit(DatastoreReaderWriter transaction, String counterId, long limit) {
+    final Key limitKey = datastore.newKeyFactory().setKind(KIND_COUNTER_LIMIT).newKey(counterId);
+
+    transaction.put(Entity.newBuilder(limitKey).set(PROPERTY_LIMIT, limit).build());
+  }
+
+  /**
+   * Reads the latest limit value from Datastore, for the specified {@param counterId}
+   */
+  public static long getLimit(Datastore datastore, String counterId) {
+    final Key limitKey = datastore.newKeyFactory().setKind(KIND_COUNTER_LIMIT).newKey(counterId);
+    final Entity limitEntity = datastore.get(limitKey);
+    if (limitEntity == null) {
+      return Long.MAX_VALUE;
+      // Or IllegalStateException("No limit found in Datastore for " + counterId);?
+    } else {
+      return limitEntity.getLong(PROPERTY_LIMIT);
+    }
+  }
+
+  /**
+   * Must be called within a TransactionCallable. Augments the transaction with certain operations
+   * that strongly consistently increment resp. decrement the counter referred to by counterId, and
+   * cause the transaction to fail to commit if the counter's associated limit is exceeded. Also
+   * spurious failures are possible.
+   *
+   * <p>Delta should be +/-1 for graceful behavior, due to how sharding is currently implemented.
+   * Updates with a larger delta are prone to spuriously fail even when the counter is not near to
+   * exceeding its limit. Failures are certain when delta >= limit / NUM_SHARDS + 1.
+   */
+  public void updateCounter(DatastoreReaderWriter transaction, String counterId, long delta) {
+    CounterSnapshot snapshot = getCounterSnapshot(counterId);
+    int shardIndex;
+    try {
+      shardIndex = snapshot.pickShardWithSpareCapacity(delta);
+    } catch (InvalidShardStateException e) {
+      try {
+        snapshot = refreshCounterSnapshot(counterId);
+        shardIndex = snapshot.pickShardWithSpareCapacity(delta);
+      } catch (InvalidShardStateException e1) {
+        LOG.warn(
+            "Invalid shard state for counter {} but cache has already been reloaded for counter. {}",
+            counterId, e1);
+        return;
+      }
+    }
+    final Key shardKey = datastore.newKeyFactory().setKind(KIND_COUNTER_SHARD)
+        .newKey(counterId + "-" + shardIndex);
+    final Entity shard = transaction.get(shardKey);
+
+    final long newCounterValue = shard.getLong(PROPERTY_SHARD_VALUE) + delta;
+    if (shard != null && Range.closed(0L, snapshot.shardCapacity(shardIndex))
+        .contains(newCounterValue) && newCounterValue <= getLimit(datastore, counterId)) {
```

Comment:
getLimit comparison seems superfluous - if it was in the range for shardCapacity, then it was less than limit by design.
Or if this was meant to catch a newly decreased limit, then the correct calculation would be a bit like that in shardCapacity... But since we aren't killing jobs when we decrease limit, I'd say this extra logic won't noticeably improve the behavior when we decrease limit anyway.

Reply:
You're right, it's completely wrong. Got thrown off by the already wrongly named `newCounterValue`